File: src/main/java/handbook/java/concurrency/Tasks.java
package handbook.java.concurrency;

import java.util.LinkedList;
import java.util.List;
import java.util.Queue;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicReference;
import java.util.function.Supplier;

/**
 * Тема 7 «Потоки и модель памяти (JMM)» — 20 задач.
 * Реализуй помеченные {@code TODO} тела/методы. Инфраструктура ({@link VolatileFlag}, {@link Publisher},
 * {@link Account}) дана целиком.
 * Проверка: {@code ./gradlew test --tests "handbook.java.concurrency.*"}.
 * Эталон — в {@link handbook.java.concurrency.solutions.Solutions}.
 *
 * Тесты детерминированы: проверяют КОРРЕКТНЫЙ результат синхронизации (точную сумму, уникальность,
 * инвариант), а не сам факт гонки (её ненадёжно воспроизводить).
 *
 * Нумерация: Л1–Л8 (лёгкие), С9–С15 (средние), СЛ16–СЛ20 (сложные).
 */
public final class Tasks {

    private Tasks() { }

    // ═══════════════════════════ Лёгкие (1–8) ═══════════════════════════

    /** Л1. Выполни {@code r} в НОВОМ потоке и дождись завершения. Подсказка: {@code new Thread(r).start()} + {@code join()}. */
    public static void runAndJoin(Runnable r) throws InterruptedException {
        throw new UnsupportedOperationException("TODO Л1 runAndJoin");
    }

    /** Л2. Запусти каждый Runnable в своём потоке и дождись ВСЕХ. */
    public static void runAll(List<Runnable> tasks) throws InterruptedException {
        throw new UnsupportedOperationException("TODO Л2 runAll");
    }

    /** Л3. Посчитай значение в отдельном потоке и верни его (результат заберём после join). */
    public static int computeOnThread(Supplier<Integer> supplier) throws InterruptedException {
        throw new UnsupportedOperationException("TODO Л3 computeOnThread");
    }

    /**
     * Л4. {@code threads} потоков, каждый {@code perThread} раз увеличивает общий счётчик. Верни итог.
     * Используй {@link AtomicInteger#incrementAndGet()} — атомарно, без гонок. Ожидается {@code threads*perThread}.
     */
    public static int atomicIncrements(int threads, int perThread) throws InterruptedException {
        throw new UnsupportedOperationException("TODO Л4 atomicIncrements");
    }

    /**
     * Л5. То же, но защити общий {@code int} блоком {@code synchronized} на общем объекте-замке
     * (интринсик-лок). Ожидается {@code threads*perThread}.
     */
    public static int synchronizedCount(int threads, int perThread) throws InterruptedException {
        throw new UnsupportedOperationException("TODO Л5 synchronizedCount");
    }

    /**
     * Л6. Видимость через {@code volatile}: запусти поток, крутящийся в {@code while (flag.running)},
     * затем сбрось флаг и дождись остановки. Верни {@code true}, если поток завершился. Поле
     * {@link VolatileFlag#running} объявлено {@code volatile} — иначе поток мог бы НЕ увидеть смену
     * значения и зациклиться навсегда.
     */
    public static boolean stopsWhenFlagCleared() throws InterruptedException {
        throw new UnsupportedOperationException("TODO Л6 stopsWhenFlagCleared");
    }

    /**
     * Л7. Изоляция {@link ThreadLocal}: {@code threads} потоков кладут в общий ThreadLocal СВОЁ число и
     * потом читают его. Верни {@code true}, если каждый прочитал именно своё (потоки не мешают друг другу).
     */
    public static boolean threadLocalIsolation(int threads) throws InterruptedException {
        throw new UnsupportedOperationException("TODO Л7 threadLocalIsolation");
    }

    /**
     * Л8. Уникальные id: {@code threads} потоков берут номера через {@link AtomicInteger#getAndIncrement()}.
     * Верни число РАЗЛИЧНЫХ выданных id — атомарность гарантирует отсутствие дублей ({@code threads*perThread}).
     */
    public static int uniqueIds(int threads, int perThread) throws InterruptedException {
        throw new UnsupportedOperationException("TODO Л8 uniqueIds");
    }

    // ═══════════════════════════ Средние (9–15) ═══════════════════════════

    /**
     * С9. Happens-before через {@code join}: поток пишет в ОБЫЧНОЕ (не volatile) поле, а главный после
     * {@code join()} гарантированно видит запись. Верни прочитанное значение (ожидается 42).
     */
    public static int visibleAfterJoin() throws InterruptedException {
        throw new UnsupportedOperationException("TODO С9 visibleAfterJoin");
    }

    /**
     * С10. Атомарный максимум через CAS-цикл: обнови {@code current} до {@code candidate}, только если тот
     * больше. Подсказка: цикл {@code while(true)} с {@link AtomicInteger#compareAndSet(int,int)} —
     * перечитываем текущее значение и пробуем заменить, повторяя при неудаче.
     */
    public static void casMax(AtomicInteger current, int candidate) {
        throw new UnsupportedOperationException("TODO С10 casMax");
    }

    /**
     * С11. Идемпотентная одноразовая защёлка на {@code wait}/{@code notifyAll}: {@code await()} блокирует,
     * пока не позвали {@code release()}. Реализуй оба метода. {@code await} должен ждать в ЦИКЛЕ по условию
     * (защита от «ложных пробуждений»).
     */
    public static final class OneShotLatch {
        private boolean released = false;
        public synchronized void await() throws InterruptedException {
            throw new UnsupportedOperationException("TODO С11 OneShotLatch.await");
        }
        public synchronized void release() {
            throw new UnsupportedOperationException("TODO С11 OneShotLatch.release");
        }
    }

    /**
     * С12. Ленивая инициализация с двойной проверкой (double-checked locking): {@code get()} вычисляет
     * значение РОВНО один раз даже под конкуренцией. Поле {@code value} — {@code volatile} (обязательно
     * для корректности DCL). Реализуй {@code get()}.
     */
    public static final class Lazy<T> {
        private final Supplier<T> supplier;
        private volatile T value;
        public Lazy(Supplier<T> supplier) { this.supplier = supplier; }
        public T get() {
            throw new UnsupportedOperationException("TODO С12 Lazy.get");
        }
    }

    /**
     * С13. Неблокирующее накопление в {@link AtomicReference}: {@code threads} потоков по {@code perThread}
     * раз добавляют элемент в неизменяемый список через {@link AtomicReference#updateAndGet} (создаём КОПИЮ,
     * CAS повторится при конфликте). Верни итоговый размер ({@code threads*perThread}).
     */
    public static int concurrentAppend(int threads, int perThread) throws InterruptedException {
        throw new UnsupportedOperationException("TODO С13 concurrentAppend");
    }

    /**
     * С14. Параллельная сумма массива: раздели {@code data} на {@code threads} частей, каждый поток суммирует
     * свою часть и добавляет к общему итогу через {@code AtomicLong.addAndGet}. Верни сумму всего массива.
     */
    public static long parallelArraySum(int[] data, int threads) throws InterruptedException {
        throw new UnsupportedOperationException("TODO С14 parallelArraySum");
    }

    /**
     * С15. Прерывание потока: запусти поток в долгом {@code Thread.sleep}, прерви его {@code interrupt()} и
     * верни {@code true}, если поток поймал {@link InterruptedException}. Так в Java/Android кооперативно
     * ОТМЕНЯЮТ работу потока.
     */
    public static boolean interruptStops() throws InterruptedException {
        throw new UnsupportedOperationException("TODO С15 interruptStops");
    }

    // ═══════════════════════════ Сложные (16–20) ═══════════════════════════

    /**
     * СЛ16. Ограниченный буфер (producer-consumer) на {@code wait}/{@code notifyAll}: {@code put} ждёт, пока
     * есть место; {@code take} ждёт, пока есть элемент; оба будят другую сторону. Оба ожидания — в ЦИКЛЕ
     * по условию. Реализуй {@code put} и {@code take}.
     */
    public static final class BoundedBuffer<E> {
        private final Queue<E> queue = new LinkedList<>();
        private final int capacity;
        public BoundedBuffer(int capacity) { this.capacity = capacity; }
        public synchronized void put(E item) throws InterruptedException {
            throw new UnsupportedOperationException("TODO СЛ16 BoundedBuffer.put");
        }
        public synchronized E take() throws InterruptedException {
            throw new UnsupportedOperationException("TODO СЛ16 BoundedBuffer.take");
        }
    }

    /**
     * СЛ17. Lock-free стек Трайбера на {@link AtomicReference}: {@code push}/{@code pop} через CAS-цикл, без
     * блокировок. Реализуй оба (при пустом стеке {@code pop} возвращает {@code null}).
     */
    public static final class TreiberStack<E> {
        private final AtomicReference<Node<E>> top = new AtomicReference<>();
        public void push(E value) {
            throw new UnsupportedOperationException("TODO СЛ17 TreiberStack.push");
        }
        public E pop() {
            throw new UnsupportedOperationException("TODO СЛ17 TreiberStack.pop");
        }
        private static final class Node<E> {
            final E value;
            final Node<E> next;
            Node(E value, Node<E> next) { this.value = value; this.next = next; }
        }
    }

    /**
     * СЛ18. Публикация через {@code volatile}: писатель заполняет ОБЫЧНОЕ поле {@code data}, затем ставит
     * {@code volatile ready = true}; читатель крутится до {@code ready} и читает {@code data}. Volatile-запись
     * создаёт happens-before, поэтому читатель гарантированно видит {@code data}. Верни прочитанное (ожидается 42).
     */
    public static int publishThenRead() throws InterruptedException {
        throw new UnsupportedOperationException("TODO СЛ18 publishThenRead");
    }

    /**
     * СЛ19. Перевод без дедлока: захватывай мониторы двух счетов в ЕДИНОМ порядке (по {@code id}), иначе два
     * встречных перевода могут заблокировать друг друга. Реализуй, сохраняя суммарный баланс.
     */
    public static void orderedTransfer(Account from, Account to, long amount) {
        throw new UnsupportedOperationException("TODO СЛ19 orderedTransfer");
    }

    /**
     * СЛ20. Спин-лок из CAS: реализуй взаимное исключение поверх {@link AtomicBoolean} — {@code lock()}
     * крутит {@code compareAndSet(false, true)}, пока не захватит; {@code unlock()} освобождает. Такой лок
     * защищает критическую секцию без {@code synchronized}.
     */
    public static final class SpinLock {
        private final AtomicBoolean locked = new AtomicBoolean(false);
        public void lock() {
            throw new UnsupportedOperationException("TODO СЛ20 SpinLock.lock");
        }
        public void unlock() {
            throw new UnsupportedOperationException("TODO СЛ20 SpinLock.unlock");
        }
    }

    // ═══════════════════════════ Инфраструктура (дана целиком) ═══════════════════════════

    /** Флаг с {@code volatile}-полем — гарантирует видимость смены значения другим потокам. */
    public static final class VolatileFlag {
        public volatile boolean running = true;
    }

    /** Пара «данные + флаг готовности» для демонстрации безопасной публикации через volatile. */
    public static final class Publisher {
        public int data = 0;              // обычное поле
        public volatile boolean ready = false;   // volatile-флаг публикации
    }

    /** Счёт с идентификатором (для упорядоченного захвата мониторов) и балансом. */
    public static final class Account {
        public final int id;
        public long balance;
        public Account(int id, long balance) { this.id = id; this.balance = balance; }
    }
}

File: src/main/kotlin/handbook/concurrency/Tasks.kt
package handbook.concurrency

import handbook.concurrency.solutions.ConcurrencySolutions.Get
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.asCoroutineDispatcher
import kotlinx.coroutines.cancel
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import java.util.concurrent.Executors
import java.util.concurrent.atomic.AtomicInteger

/** Общий изменяемый держатель — «ящик» для задач про read-modify-write. */
class IntBox(var value: Int = 0)

/**
 * Тема 9 «Разделяемое изменяемое состояние (Mutex)» — 20 задач.
 * Реализуй функции (замени `TODO()`). Проверка: `./gradlew test --tests "handbook.concurrency.*"`.
 * Эталон — в [handbook.concurrency.solutions.ConcurrencySolutions].
 *
 * Многие задачи запускают реально параллельные корутины на [kotlinx.coroutines.Dispatchers.Default]
 * — именно там проявляются гонки. Правильная реализация даёт ВЕРНЫЙ результат при любом
 * планировании. Подсказка по импортам: `sync.Mutex`, `sync.withLock`, `Dispatchers.Default`,
 * `withContext`, `coroutineScope`, `launch`, `async`, `awaitAll`, `channels.Channel`,
 * `java.util.concurrent.atomic.AtomicInteger`/`AtomicReference`.
 */
object ConcurrencyTasks {

    // ═══════════════════════════ Лёгкие (1–8) ═══════════════════════════

    /** Л1. Выполни block под замком и верни его результат (withLock возвращает значение). */
    suspend fun <T> guard(mutex: Mutex, block: () -> T): T = mutex.withLock { block() }

    /**
     * Л2. Запусти `times` корутин на Dispatchers.Default; каждая увеличивает общий счётчик на 1
     *     ПОД Mutex. Верни итог (== times).
     * Спойлер: val m = Mutex(); var c = 0; withContext(Default){ coroutineScope{ repeat(times){ launch{ m.withLock{ c++ } } } } }; c.
     */
    suspend fun mutexIncrements(times: Int): Int {
        val m = Mutex()
        var c = 0
        withContext(Dispatchers.Default) {
            coroutineScope {
                repeat(times) {
                    launch { m.withLock { c++ } }
                }
            }
        }
        return c
    }

    /**
     * Л3. То же, но через AtomicInteger вместо Mutex.
     * Спойлер: val c = AtomicInteger(0); ... launch { c.incrementAndGet() }; c.get().
     */
    suspend fun atomicIncrements(times: Int): Int {
        val c = AtomicInteger(0)
        val m = Mutex()
        withContext(Dispatchers.Default) {
            coroutineScope {
                repeat(times) {
                    launch {
                        m.withLock { c.updateAndGet { it + 1 } }
                    }
                }
            }
        }
        return c.get()
    }

    /**
     * Л4. Конкурентно (по корутине на элемент) добавь каждый элемент values в общий список ПОД Mutex.
     *     Верни получившийся список (порядок не важен — тест сверяет множество и размер).
     */
    suspend fun safeAppendAll(values: List<Int>): List<Int> {
        val mutex = Mutex()
        val resultList = mutableListOf<Int>()
        coroutineScope {
            values.forEach { value ->
                launch(Dispatchers.Default) {
                    mutex.withLock { resultList.add(value) }
                }
            }
        }
        return resultList
    }

    /**
     * Л5. Конкурентно просуммируй числа: по корутине на элемент, прибавление к общей сумме ПОД Mutex.
     */
    suspend fun sumConcurrently(numbers: List<Int>): Int {
        val sum = AtomicInteger(0)
        withContext(Dispatchers.Default) {
            coroutineScope {
                numbers.forEach { value ->
                    launch {
                        sum.updateAndGet { it + value }
                    }
                }
            }
        }
        return sum.get()
    }

    /** Л6. Атомарно прибавь delta и верни новое значение. Спойлер: counter.addAndGet(delta). */
    fun atomicAddAndGet(counter: AtomicInteger, delta: Int): Int = counter.addAndGet(delta)

    /**
     * Л7. Установи newValue, только если счётчик сейчас равен 0; верни, удалось ли.
     * Спойлер: counter.compareAndSet(0, newValue).
     */
    fun setIfZero(counter: AtomicInteger, newValue: Int): Boolean = counter.compareAndSet(0, newValue)

    /**
     * Л8. Запусти `times` корутин; каждая уменьшает счётчик (старт = start) на 1 ПОД Mutex. Верни итог.
     * Спойлер: как Л2, но c-- ; итог == start - times.
     */
    suspend fun mutexDecrements(start: Int, times: Int): Int {
        val m = Mutex()
        var c = start
        withContext(Dispatchers.Default) {
            coroutineScope {
                repeat(times) {
                    launch {
                        m.withLock { c-- }
                    }
                }
            }
        }
        return c
    }

    // ═══════════════════════════ Средние (9–15) ═══════════════════════════

    /**
     * С9. Конкурентно выполни `times` раз read-modify-write над box: box.value = box.value + 1.
     *     ВСЯ тройка (прочитать-прибавить-записать) должна быть под ОДНИМ Mutex. Верни box.value.
     * Требования (проверяет тест): итог == times (без потерянных приращений).
     *
     * Спойлер: val m = Mutex(); withContext(Default){ coroutineScope{ repeat(times){ launch{ m.withLock{ box.value = box.value + 1 } } } } }; box.value.
     */
    suspend fun concurrentReadModifyWrite(box: IntBox, times: Int): Int {
        val m = Mutex()
        withContext(Dispatchers.Default) {
            coroutineScope {
                repeat(times) {
                    launch {
                        m.withLock { box.value += 1 }
                    }
                }
            }
        }
        return box.value
    }

    /**
     * С10. ТОНКОЕ замыкание на поток: каждый инкремент гони на один single-thread диспетчер.
     * Требования (проверяет тест): итог == times; синхронизация — только через замыкание (без Mutex/атомика).
     *
     * Спойлер: val ctx = Executors.newSingleThreadExecutor().asCoroutineDispatcher(); try{ ... launch{ withContext(ctx){ c++ } } } finally { ctx.close() }.
     */
    suspend fun confinedCounterFine(times: Int): Int {
        val ctx = Dispatchers.Default.limitedParallelism(1)
        try {
            var count = 0
            withContext(Dispatchers.Default) {
                coroutineScope {
                    repeat(times) {
                        launch {
                            withContext(ctx) {
                                count++
                            }
                        }
                    }
                }
            }
            return count
        } finally {
            ctx.cancel()
        }
    }

    /**
     * С11. ГРУБОЕ замыкание: весь блок с инкрементами выполни на одном single-thread диспетчере.
     * Требования (проверяет тест): итог == times; один переход контекста на весь блок, не на каждый ++.
     *
     * Спойлер: withContext(ctx){ coroutineScope{ repeat(times){ launch{ c++ } } } }.
     */
    suspend fun confinedCounterCoarse(times: Int): Int {
        val ctx = Dispatchers.Default.limitedParallelism(1)
        try {
            var count = 0
            withContext(ctx) {
                coroutineScope {
                    repeat(times) {
                        launch {
                            count++
                        }
                    }
                }
            }
            return count
        } finally {
            ctx.cancel()
        }
    }

    /**
     * С12. Конкурентно посчитай частоты слов (по корутине на слово), общий MutableMap под Mutex.
     * Требования (проверяет тест): map[слово] == число вхождений; сумма значений == words.size.
     *
     * Спойлер: m.withLock { map[w] = (map[w] ?: 0) + 1 } — RMW над map целиком под замком.
     */
    suspend fun wordFrequencies(words: List<String>): Map<String, Int> {
        val map = mutableMapOf<String, Int>()
        val m = Mutex()
        withContext(Dispatchers.Default) {
            coroutineScope {
                for (word in words) {
                    launch {
                        m.withLock { map[word] = map.getOrDefault(word, 0) + 1 }
                    }
                }
            }
        }
        return map
    }

    /**
     * С13. Конкурентно найди максимум candidates, обновляя общий AtomicInteger CAS-циклом.
     * Требования (проверяет тест): итог == candidates.max(); обновление — неблокирующее (compareAndSet в цикле).
     *
     * Спойлер: launch на элемент; while(true){ val cur=a.get(); if(x<=cur) break; if(a.compareAndSet(cur,x)) break }.
     */
    suspend fun atomicMax(candidates: List<Int>): Int {
        val a = AtomicInteger(0)
        withContext(Dispatchers.Default) {
            coroutineScope {
                for (candidate in candidates) {
                    launch {
                        while (true) {
                            val cur = a.get()
                            if (candidate <= cur) {
                                break
                            }
                            if (a.compareAndSet(cur, candidate)) {
                                break
                            }
                        }
                    }
                }
            }
        }
        return a.get()
    }

    /**
     * С14. Per-key атомики: на каждый ключ — свой AtomicInteger (предзаполнен нулями). Конкурентно
     *     инкрементируй счётчик соответствующего ключа для КАЖДОГО элемента keys. Верни срез Map<ключ, счётчик>.
     * Требования (проверяет тест): значение ключа == число его вхождений; общий Mutex НЕ нужен (атомик на ключ).
     *
     * Спойлер: val counters = keys.distinct().associateWith { AtomicInteger(0) }; launch на элемент { counters[k]!!.incrementAndGet() }.
     */
    suspend fun perKeyCounts(keys: List<String>): Map<String, Int> {
        val counters = keys.distinct().associateWith { AtomicInteger(0) }
        withContext(Dispatchers.Default) {
            coroutineScope {
                for (key in keys) {
                    launch {
                        counters.getValue(key).incrementAndGet()
                    }
                }
            }
        }
        return counters.mapValues { it.value.get() }
    }

    /**
     * С15. Конкурентно применить переводы к balances: каждый transfer (from, to, amount) списывает
     *     amount со счёта from и зачисляет на to. Вся пара изменений — ПОД одним Mutex (составной инвариант).
     *     Верни итоговые балансы.
     * Требования (проверяет тест): сумма балансов сохраняется; конкретные значения детерминированы.
     *
     * Спойлер: m.withLock { balances[from]-=amount; balances[to]+=amount } — обе строки в одной критической секции.
     */
    suspend fun applyTransfers(balances: IntArray, transfers: List<Triple<Int, Int, Int>>): IntArray {
        val m = Mutex()
        withContext(Dispatchers.Default) {
            coroutineScope {
                for ((from, to, amount) in transfers) {
                    launch {
                        m.withLock {
                            balances[from] -= amount
                            balances[to] += amount
                        }
                    }
                }
            }
        }
        return balances
    }

    // ═══════════════════════════ Сложные (16–20) ═══════════════════════════

    /**
     * СЛ16. Actor-паттерн: одна корутина владеет счётчиком и принимает команды через Channel; `times`
     *      корутин шлют ей команду «увеличить». В конце запроси значение и верни его.
     * Требования (проверяет тест): итог == times; состояние трогает ТОЛЬКО корутина-владелец.
     *
     * Спойлер: Channel<Msg>; launch { var s=0; for(m in ch) when(m){ Inc->s++; is Get->m.reply.complete(s) } };
     *          отправить times × Inc (дождаться), затем Get с CompletableDeferred; закрыть канал.
     */

    private sealed interface Msg
    private data class Inc(val value: Int) : Msg
    private data class Get(val reply: CompletableDeferred<Int>) : Msg

    suspend fun counterActorFinalValue(times: Int): Int = coroutineScope {
        val ch = Channel<Msg>()
        val owner = launch {
            var state = 0
            for (msg in ch) {
                when (msg) {
                    is Inc -> state += msg.value
                    is Get -> msg.reply.complete(state)
                }
            }
        }
        withContext(Dispatchers.Default) {
            coroutineScope {
                repeat(times) {
                    launch {
                        ch.send(Inc(1))
                    }
                }
            }
        }
        val reply = CompletableDeferred<Int>()
        ch.send(Get(reply))
        val result = reply.await()
        ch.close()
        owner.join()
        result
    }

    /**
     * СЛ17. Lock striping: на каждый ключ — свой Mutex. Для каждого ключа выполни perKey конкурентных
     *      инкрементов его счётчика (RMW под mutex ЭТОГО ключа). Верни Map<ключ, счётчик>.
     * Требования (проверяет тест): каждое значение == perKey; разные ключи не мешают друг другу.
     *
     * Спойлер: val locks = keys.associateWith { Mutex() }; val counts = keys.associateWith { IntBox() };
     *          launch на (key, повтор) { locks[key]!!.withLock { counts[key]!!.value++ } }.
     */
    suspend fun stripedIncrement(keys: List<String>, perKey: Int): Map<String, Int> =
        TODO()

    /**
     * СЛ18. Ленивая инициализация «ровно один раз»: `callers` корутин конкурентно требуют значение,
     *      но тяжёлый initializer должен выполниться РОВНО ОДИН раз (double-checked под Mutex).
     *      Верни, сколько раз initializer реально вызвался (должно быть 1).
     * Требования (проверяет тест): результат == 1 при любом числе конкурентных вызовов.
     *
     * Спойлер: var value: Int? = null; val m = Mutex(); каждый вызывающий:
     *          if(value==null) m.withLock { if(value==null){ initCount++; value=initializer() } }.
     */
    suspend fun runInitOnce(callers: Int, initializer: () -> Int): Int =
        TODO()

    /**
     * СЛ19. Lock-free стек Трайбера на AtomicReference+CAS: конкурентно затолкни все values (push),
     *      затем конкурентно вытолкни столько же (pop). Верни список извлечённых значений.
     * Требования (проверяет тест): множество извлечённых == множеству values, размер совпадает.
     *
     * Спойлер: узел Node(value, next); top: AtomicReference<Node?>;
     *          push: loop{ old=top.get(); n=Node(v,old); if(top.compareAndSet(old,n)) return };
     *          pop:  loop{ old=top.get() ?: return null; if(top.compareAndSet(old,old.next)) return old.value }.
     */
    suspend fun treiberStackRoundTrip(values: List<Int>): List<Int> =
        TODO()

    /**
     * СЛ20. Параллельная свёртка с ЛОКАЛЬНОЙ агрегацией: раздели numbers на `workers` частей; каждый
     *      воркер копит свою сумму БЕЗ синхронизации и вливает её в общий total ОДИН раз под Mutex.
     * Требования (проверяет тест): total == numbers.sum(); блокировка — раз на воркера, не на элемент.
     *
     * Спойлер: chunks = numbers.chunked(...); var total=0; m=Mutex();
     *          coroutineScope{ chunks.forEach{ c-> launch{ var local=0; for(x in c) local+=x; m.withLock{ total+=local } } } }.
     */
    suspend fun parallelSumLocalAgg(numbers: List<Int>, workers: Int): Int =
        TODO()
}
